src/ring_theory/adjoin/basic.lean

```diff
@@ -110,6 +118,72 @@ lemma adjoint_prod_le (s : set A) (t : set B) :
   adjoin R (set.prod s t) ≤ (adjoin R s).prod (adjoin R t) :=
 adjoin_le $ set.prod_mono subset_adjoin subset_adjoin
 
+lemma adjoin_le_prod (s) (t) : adjoin R ((set.insert (1, 0) ((linear_map.inl R A B) '' s)) ∪
+  (set.insert (0, 1) ((linear_map.inr R A B) '' t))) ≤ (adjoin R s).prod (adjoin R t) :=
```

Comment:
I think it would make sense to define:
```
def _root_.alg_hom.fst : A × B →ₐ[R] A :=
{ commutes' := λ r, rfl, .. ring_hom.fst A B}

def _root_.alg_hom.snd : A × B →ₐ[R] A :=
{ commutes' := λ r, rfl, .. ring_hom.fst A B}
```
along with the `inr` and `inl` maps. These probably belong in `algebra/algebra/basic`.

Reply:
I 100% agree with `_root_.alg_hom.fst`, but there is no way to produce the maps the other direction, they are not ring hom...